### PR TITLE
refactor: CC => FC

### PR DESCRIPTION
**Pull request by li-jia-nan**
[[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md)]

### 🤔 This is a ...

- [ ] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [x] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | refactor: CC => FC |
| 🇨🇳 Chinese | refactor: CC => FC |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] TypeScript definition is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

File: components/locale-provider/index.tsx [hunk 1]
```diff
@@ -48,51 +47,37 @@ export interface Locale {
 }
 
 export interface LocaleProviderProps {
-  locale: Locale;
+  locale?: Locale;
```
Comment (NWYLZW) [inline components/locale-provider/index.tsx:50]: 这个地方应该不用 optional，你在下面的 line 57 provider 已经保证了一定存在这个 props 了

File: components/locale-provider/index.tsx [hunk 2]
```diff
@@ -54,45 +53,33 @@ export interface LocaleProviderProps {
...
+    };
+  }, [locale]);
 
-  componentDidUpdate(prevProps: LocaleProviderProps) {
-    const { locale } = this.props;
-    if (prevProps.locale !== locale) {
-      changeConfirmLocale(locale && locale.Modal);
-    }
-  }
+  const getMemoizedContextValue = memoizeOne(
```
Comment (zombieJ) [inline components/locale-provider/index.tsx:74]: 这个看看能不能改成 useMemo？

File: components/locale-provider/LocaleReceiver.tsx [hunk 3]
```diff
@@ -7,67 +7,58 @@ import defaultLocaleData from './default';
...
-  context: LocaleContextProps;
-
-  getLocale(): NonNullable<Locale[C]> {
-    const { componentName, defaultLocale } = this.props;
-    const locale = defaultLocale || defaultLocaleData[componentName ?? 'global'];
-    const antLocale = this.context;
-    const localeFromContext = componentName && antLocale ? antLocale[componentName] : {};
+  const getLocale = React.useMemo<NonNullable<Locale[C]>>(() => {
+    const locale = defaultLocale || defaultLocaleData[componentName];
+    const localeFromContext = antLocale ? antLocale[componentName] : {};
```
Comment (zombieJ) [inline components/locale-provider/LocaleReceiver.tsx:27]: 这个和之前的不等价，可以改成：
```tsx
const localeFromContext = antLocale?.[componentName] ?? {};
```

File: components/locale-provider/LocaleReceiver.tsx [hunk 4]
```diff
@@ -7,67 +7,58 @@ import defaultLocaleData from './default';
...
+  defaultLocale?: Locale[C] | (() => Locale[C]),
+): [Locale[C]] => {
+  const antLocale = React.useContext<LocaleContextProps | undefined>(LocaleContext);
 
+  const getLocale = React.useMemo<NonNullable<Locale[C]>>(() => {
+    const locale = defaultLocale || defaultLocaleData[componentName];
+    const localeFromContext = antLocale ? antLocale[componentName] : {};
     return {
-      ...(typeof locale === 'function' ? (locale as Function)() : locale),
+      ...(locale instanceof Function ? locale() : locale),
```
Comment (zombieJ) [inline components/locale-provider/LocaleReceiver.tsx:58]: 这里不用改，还是 typeof 即可。

File: components/locale-provider/index.tsx (hunk 2 above)
Comment (li-jia-nan) [reply components/locale-provider/index.tsx:74]: 这个是函数，应该用useCallback好一点吧

File: components/locale-provider/LocaleReceiver.tsx (hunk 4 above)
Comment (li-jia-nan) [reply components/locale-provider/LocaleReceiver.tsx:58]: 这里是为了和上面的保持一致，两个地方逻辑是一模一样的，不过没啥影响，我改回去

File: components/locale-provider/LocaleReceiver.tsx (hunk 3 above)
Comment (li-jia-nan) [reply components/locale-provider/LocaleReceiver.tsx:27]: done

File: components/locale-provider/__tests__/cached-context.test.tsx [hunk 5]
```diff
@@ -1,51 +1,37 @@
-import React, { memo, useContext, useRef, useState } from 'react';
-import { render, fireEvent } from '../../../tests/utils';
-import LocaleProvider, { ANT_MARK } from '..';
+import React, { memo, useContext } from 'react';
+import { fireEvent, pureRender } from '../../../tests/utils';
+import LocaleProvider from '..';
 import LocaleContext from '../context';
 
-const defaultLocale = { locale: 'locale' };
+let [innerCount, outerCount] = [0, 0];
```
Comment (zombieJ) [inline components/locale-provider/__tests__/cached-context.test.tsx:6]: 这个看起来有点怪异，没必要解构。直接申明 2 个变量即可。

**Review by zombieJ — APPROVED** [review]

**Pull request merged**